Add instructions to configure HPA on openshift

davidor opened the pull request:
This PR adds a document that explains how to setup Horizontal pod Autoscaling (HPA) in Openshift.

andrewdavidmackenzie commented on docs/openshift_horizontal_scaling.md:
````diff
@@ -0,0 +1,105 @@
+```bash
+oc describe hpa backend-listener
+oc describe hpa backend-worker
+```
````
Add a little bit about seeing it work when testing, load increases?

davidor commented on docs/openshift_horizontal_scaling.md:
```diff
@@ -0,0 +1,105 @@
+
+- Set the `CONFIG_REDIS_ASYNC` to true in the listener and worker deployment configs.
+- Configure listeners to use the Falcon web-server instead of Puma. In order to
+do this, modify the command in the listener deployment config. Replace this:
```
I agree. In the future, this should be implemented in the operator, but I wanted to document all the apisonator-specific knowledge here and validate the idea first.

davidor commented on docs/openshift_horizontal_scaling.md:
```diff
@@ -0,0 +1,105 @@
+DB and the data DB in the same Redis process under different DB indexes. This is
+not supported by the async redis library, so we need to either place both DBs in
+the same Redis process or use two separate ones. To configure this, change the
+`CONFIG_REDIS_PROXY` and `REDIS_QUEUES_URL` envs in the listener and worker
```
This variable was introduced a long time ago. We can discuss ENV name changes in a separate PR because they require other changes to ensure backwards compatibility.

davidor commented on docs/openshift_horizontal_scaling.md:
```diff
@@ -0,0 +1,105 @@
+### Set up horizontal pod autoscaling
+
+We need to decide the target CPU. It is defined as a percentage of the requested
+CPU. In a default installation, the backend-listener pod requests 0.5 CPUs. In
```
I added some changes to clarify this.

unleashed commented on docs/openshift_horizontal_scaling.md:
```diff
@@ -0,0 +1,105 @@
+# Openshift Horizontal Pod Autoscaling (HPA)
+
+It is possible to configure Apisonator to auto scale when deployed in Openshift.
```
nit:
```suggestion
It is possible to configure Apisonator to automatically scale when deployed on Openshift.
```

unleashed replied:
perhaps mention this can also be done in bare k8s?

davidor replied:
Done :+1:

unleashed commented on docs/openshift_horizontal_scaling.md:
````diff
@@ -0,0 +1,105 @@
+        - '-x'
+        - /dev/stdout
+```
+- Set the `PUMA_WORKERS` env to 1 in the listener deployment. This will deploy 1
````
How about introducing `LISTENER_WORKERS`, look it up with higher precedence in the code, and if not present fall back to `PUMA_WORKERS`?

unleashed commented on docs/openshift_horizontal_scaling.md:
```diff
@@ -0,0 +1,105 @@
+### Enable async
+
+- Set the `CONFIG_REDIS_ASYNC` to true in the listener and worker deployment configs.
+- Configure listeners to use the Falcon web-server instead of Puma. In order to
```
This depends on #150.

davidor replied:
Yes, go approve it :smile:

unleashed commented on docs/openshift_horizontal_scaling.md:
```diff
@@ -0,0 +1,105 @@
+listener. In a default installation, the backend-worker pod requests 0.150 CPU
+cores. Following the same logic, we need to define a target CPU of 533
+(0.8/0.150 = 5.33). Pay attention to the CPU request of each pod, as they might
+change between versions, and adapt the target CPU % values accordingly.
```
So here comes the logical question: why would we want to scale so much more the workers CPU demand (to 0.8) when we only request initially 0.150 millicores (when compared to the listeners which start way higher)? We could add something, even if a little vague, about why this is the case.

davidor replied:
Even if you request only 0.150, you can ask for a limit of 1.

davidor replied:
In my opinion, it's a bit confusing to define the target CPU as a function of the pod request. It'd be more intuitive to just define the number of cores.

unleashed replied:
And the docs are not helping. Will look them up in k8s docs and see if there is any clearer explanation.

davidor commented:
@unleashed I added `LISTENER_WORKERS` in the other PR (#150) as you suggested.

davidor commented:
@andrewdavidmackenzie I've addressed your comments.

@unleashed I have replaced `PUMA_WORKERS` with `LISTENER_WORKERS`. Regarding `--cpu-percent`, I prefer to leave it as it is. If the reader wants to know more about that parameter, there are links to the kubernetes and openshift docs that he can check.

andrewdavidmackenzie commented:
Great! Go for it.
